src/System.Data.SqlClient/src/System/Data/SqlClient/SqlStatistics.cs

```diff
@@ -218,6 +221,114 @@ internal void UpdateStatistics()
                 _connectionTime = long.MaxValue;
             }
         }
+
+        // We subclass Dictionary to provide our own implementation of CopyTo, Keys.CopyTo, and
+        // Values.CopyTo to match the behavior of Hashtable, which is used in the full framework:
+        //
+        //  - When arrayIndex > array.Length, Hashtable throws ArgumentException whereas Dictionary
+        //    throws ArgumentOutOfRangeException.
+        //
+        //  - Hashtable specifies the ArgumentOutOfRangeException paramName as "arrayIndex" whereas
+        //    Dictionary uses "index".
+        //
+        //  - When the array is of a mismatched type, Hashtable throws InvalidCastException whereas
+        //    Dictionary throws ArrayTypeMismatchException.
+        //
+        //  - Hashtable allows copying values to a long[] array via Values.CopyTo, whereas Dictionary
+        //    throws ArgumentException due to the "Target array type is not compatible with type of
+        //    items in the collection" (when Dictionary<object, object> is used).
+        //
+        // Ideally this would derive from Dictionary<string, long>, but that would break compatibility
+        // with the full framework, which allows adding keys/values of any type.
```

Comment:
Nice.  I received an email from you with some comment reply. However I am not able to navigate to the comment in the web view.
I see your point about the subclass and the difference in behavior. Since this bubbles up all the way to the contract and a public API, it makes sense to imitate the Hashtable behavior.